### PR TITLE
Modifier to apply energy corrections to photons and electrons (miniAOD) for 74X

### PR DESCRIPTION
**Pull request by lgray**
Same as #10690.

ESPrefers are current used to get the necessary tags.

They will removed once the weights will be integrated into the GT.

Working on comparison plots for the resolution.

Supersedes #10613. (had to remove residual .root files in the history)

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@diguida @mmusich Also blocked by not having the GBRD records in the GT. Please let me know when they are available! Thanks!

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
will fix.

**Comment by Dr15Jones** [comment]
```std::make_unique<>``` is a C++14 feature and, unfortunately, CMSSW_7_4 only works up to C++11. I got burned by that this week as well. :)

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
nice.

**Comment by mmusich** [comment]
please test

**Comment by lgray** [comment]
@davidlange6 These DAS problems are becoming irritatingly bad.

**Comment by davidlange6** [comment]
becoming? they are well past that for me..

**Comment by lgray** [comment]
@davidlange6 Please excuse my understatement of the issue. It would have been unwise to post what I actually want to say.

**Comment by lgray** [comment]
@mmusich possible to get a +1 here for the advancement of the global tag? thanks!

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py [hunk 1]
```diff
@@ -163,9 +163,44 @@ def miniAOD_customizeCommon(process):
         cms.InputTag("reducedEgamma","reducedEERecHits")
     process.electronIDValueMapProducer.esReducedRecHitCollection = \
         cms.InputTag("reducedEgamma","reducedESRecHits")
+    process.electronMVAValueMapProducer.src = \
+        cms.InputTag("reducedEgamma","reducedGedGsfElectrons")
+    process.electronRegressionValueMapProducer.src = \
+        cms.InputTag("reducedEgamma","reducedGedGsfElectrons")
+    process.electronRegressionValueMapProducer.ebReducedRecHitCollection = \
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: these modifications are not present in 75X/76X versions. Why needed here?

Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: ... I guess that's interference with #10598, which is not in this PR.

File: PhysicsTools/PatAlgos/python/slimming/modifyPrimaryPhysicsObjects_cff.py [hunk 2]
```diff
@@ -22,6 +24,7 @@
 
 modifyPrimaryPhysicsObjects = cms.Sequence( electronMVAValueMapProducer *
                                             photonIDValueMapProducer * photonMVAValueMapProducer * 
+                                            electronRegressionValueMapProducer * photonRegressionValueMapProducer *
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/modifyPrimaryPhysicsObjects_cff.py]: these are not in 75X/76X
why needed here?
Isn't the modification in egmGsfElectronIDSequence enough?

**Comment by slava77** [comment]
compared to #10689 or #10690 
- PhysicsTools/PatAlgos/plugins/PATElectronSlimmer.cc and PhysicsTools/PatAlgos/plugins/PATPhotonSlimmer.cc have setEventContent called in produce method https://github.com/cms-sw/cmssw/pull/10690/files#diff-d623ac246962e988e6fc41f346248158L92
- process.electronRegressionValueMapProducer.ebReducedRecHitCollection and process.photonRegressionValueMapProducer.ebReducedRecHitCollection (similar for ee and es) in miniAOD_tools.py are modified only in 74X version https://github.com/cms-sw/cmssw/pull/10693/files#r37719611
- changes in  PhysicsTools/PatAlgos/python/slimming/modifyPrimaryPhysicsObjects_cff.py  are not present in 75X/76X (these are additions of electronRegressionValueMapProducer  and photonRegressionValueMapProducer )

... not surprisingly, #10598 doesn't merge on top of this PR.
I didn't try the other way around, but from the context, I'm guessing it wouldn't work either.
What to do?

File: PhysicsTools/PatAlgos/python/slimming/modifyPrimaryPhysicsObjects_cff.py (hunk 2 above)
Comment (lgray) [inline PhysicsTools/PatAlgos/python/slimming/modifyPrimaryPhysicsObjects_cff.py]: This sequence isn't used in MiniAOD (but only in MiniAOD-from-MiniAOD,
which was never fully setup in 75X/76X).
That's why these differences are here.

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py (hunk 1 above)
Comment (lgray) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: The rechits collections you point at don't matter so much.
So this is OK, but it is indeed likely causing conflicts.

**Comment by lgray** [comment]
@slava77 First bullet is a good catch. I'll make sure that change gets into 74X.

Suggest that I fix that point then rebase ontop of #10598, or #10598 on top of this. Then we either merge separately or have a new PR that supersedes.

**Comment by lgray** [comment]
Since #10598 is already signed, I will rebase this PR on top of it.

**Comment by lgray** [comment]
@gpetruc for your PR of PRs, once I am done here merge #10598 first and then this PR. I'll ping you when I am finished.

**Comment by gpetruc** [comment]
I have already merged it and fixed the merge in
https://github.com/gpetruc/cmssw/tree/MiniAODSpring15pass2-from74X-dev-from1.11
( which merges everything up to v1.11-d3 in
https://twiki.cern.ch/twiki/bin/view/CMS/MiniAODSpring15pass2 )
So, I'd prefer if you could just merge #10598 and fix the conflict in the
merge commit.

Giovanni

**Comment by lgray** [comment]
As commented by gchat, I'd prefer to keep the commit history linear and reorganize the PR of PRs such that it merges the appropriate things.

Either option works for me, I just find one a bit more clean in the end if we need to emergency-revert something.

Perhaps @davidlange6, @cvuosalo, or @slava77 have something to say here.

**Comment by gpetruc** [comment]
Irrespectively of whether you merge #10598 into this or you rebase this on
top of #10598, this PR will end up containing both features, so I don't see
the difference

**Comment by lgray** [comment]
Right but if you rebase then you will not have a pointer in this PR that
says "I merged this other PR" and so if you revert a change you might
remove both.
If you merge a the PR you rebased on before this one then you have one
merged PR corresponding to each set of given features (rather than one PR
with a set of features, and other PR that points to two sets of features
and has a fix for one of them included).

It is just a matter of being somewhat less confusing in the history.

**Comment by lgray** [comment]
Also, I have to add some things to this PR *anyway* because some things didn't make it in the backport I originally received. You'll have to play with the PR of PRs in any case.

**Comment by gpetruc** [comment]
Ok then I'll drop #10693 from the things I'm testing and I won't test it
anymore until it's fully signed and frozen

**Comment by lgray** [comment]
OK, after at the conflicts, I am going to have to merge in #10598 to the latest IB, then base this branch after that merge commit. Should be clean enough. 

¯\_(ツ)_/¯

**Comment by lgray** [comment]
@cmsbuild please test
Short matrix checks out.

**Comment by slava77** [comment]
@lgray are you done with commits?
I see one extra after you triggered tests.

File: PhysicsTools/PatAlgos/python/slimming/modifyPrimaryPhysicsObjects_cff.py (hunk 2 above)
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/modifyPrimaryPhysicsObjects_cff.py]: On 8/24/15 2:12 AM, Lindsey Gray wrote:
> In
> PhysicsTools/PatAlgos/python/slimming/modifyPrimaryPhysicsObjects_cff.py
> <https://github.com/cms-sw/cmssw/pull/10693#discussion_r37724172>:
> 
>> @@ -22,6 +24,7 @@
>>  
>>  modifyPrimaryPhysicsObjects = cms.Sequence( electronMVAValueMapProducer *
>>                                              photonIDValueMapProducer * photonMVAValueMapProducer * 
>> +                                            electronRegressionValueMapProducer * photonRegressionValueMapProducer *
> 
> This sequence isn't used in MiniAOD (but only in MiniAOD-from-MiniAOD,
> which was never fully setup in 75X/76X). That's why these differences
> are here.


So, do we need a follow-up PR in 75X/76X to have the
MiniAOD-from-MiniAOD working there?

The timescale of moving to 75X is not that long.

**Comment by lgray** [comment]
@slava77 It appears there was some miscommunication with the regression developers when the original PR was minted. They've also been on vacation so just now getting confirmations and double checks.

**Comment by lgray** [comment]
@slava77 For instance, I got the mail from Benoit that everything was in fractionized shower variables from Benoit only today. Then noticed that there were other places where full5x5 was hardcoded.

**Comment by lgray** [comment]
@cmsbuild please test

@slava77 I am done double checking for residual use of full5x5. Done messing with this PR.

**Comment by lgray** [comment]
@slava77 also, looks like jenkins died here or something else weird happened.

**Comment by slava77** [comment]
visually this looks consistent with what was added to 76X.
So, looks like it's going to go smoothly.

@mmusich @diguida 
Are there any issues with the GT? #10883 is still not merged nor signed by you.

**Comment by lgray** [comment]
Yes this is a straight backport from 76X now, as opposed to what was going
on before. Things are consistent by construction.

(Sent from my Nexus 6)

**Comment by mmusich** [comment]
+1
@slava77 nothing wrong with the GT (actually waiting for your answer on that thread...)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10693	f94a1dc

- code changes are now consistent with what was applied in 76X (on top of #10598): changes from pull #10689 as of e9ad0ce (merged) and pull #10921 0c57abf (signed) all show up here. Changes in the code behavior are in line with a combination of the corresponding PRs.
- jenkins tests pass and comparisons with the baseline show differences
    - L1 plots are showing some non-trivial features https://github.com/cms-sw/cmssw/pull/10883#issuecomment-134367516 .. that's unrelated to this PR
    - electron miniAOD DQM shows changes in the energy, expected from the applied regressions corrections
    - reco products are unchanged
- checked locally in CMSSW_7_4_10_patch1 /test area sign580/, comparing on top of #10883 
    - changes are only in the electron miniAOD DQM plots
    - product size changes are only in MINIAOD  in slimmedElectrons and slimmedPhotons (in fact, the size change is coming from #10598 , the ID changes).
    - prompt-reco like process allocated memory increases by about 25 MB, expected increase from the regression GBRForestD
    - timing changes are fairly minor (although photonIDValueMapProducer could've been made faster as noticed in https://github.com/cms-sw/cmssw/pull/10598#issuecomment-134015166
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.01%         0.00 ms/ev ->         1.09 ms/ev photonMVAValueMapProducer
       added      +0.00%         0.00 ms/ev ->         0.52 ms/ev photonRegressionValueMapProducer
       added      +0.17%         0.00 ms/ev ->        24.19 ms/ev photonIDValueMapProducer
       added      +0.00%         0.00 ms/ev ->         0.70 ms/ev electronRegressionValueMapProducer
       added      +0.02%         0.00 ms/ev ->         2.57 ms/ev egmPhotonIDs
       added      +0.00%         0.00 ms/ev ->         0.44 ms/ev electronMVAValueMapProducer
   +1.153350      +0.01%         0.46 ms/ev ->         1.72 ms/ev slimmedPhotons
   +0.453396      +0.00%         0.49 ms/ev ->         0.78 ms/ev slimmedElectrons
   +0.298421      +0.01%         2.77 ms/ev ->         3.74 ms/ev egmGsfElectronIDs
   +0.218959      +0.00%         0.78 ms/ev ->         0.98 ms/ev patPhotons

```

**Comment by slava77** [comment]
@monttj 
please check this one

**Comment by monttj** [comment]
+1

**Comment by slava77** [comment]
@davidlange6 this PR is merged now in #10649

**Comment by davidlange6** [comment]
Closing PRs integrated in #10649

**Pull request closed without merging**